internal/client/client_batch.go

```diff
@@ -604,23 +604,50 @@ func (c *batchCommandsClient) send(forwardedHost string, req *tikvpb.BatchComman
 			zap.Uint64s("requestIDs", req.RequestIds),
 			zap.Error(err),
 		)
-		c.failPendingRequests(err)
+		c.fastFailRequests(err, req.RequestIds)
 	}
 }
 
 // `failPendingRequests` must be called in locked contexts in order to avoid double closing channels.
-func (c *batchCommandsClient) failPendingRequests(err error) {
+// when enable-forwarding is true, the `forwardedHost` maybe not empty.
+// failPendingRequests fails all pending requests which req.forwardedHost equals to forwardedHost parameter.
+// Why need check `forwardedHost`? Here is an example, when enable-forwarding is true, and this client has network issue with store1:
+//   - some requests are sent to store1 with forwarding, such as forwardedHost is store2, those requests will success.
+//   - some requests are sent to store1 without forwarding, and maybe failed then call failPendingRequests,
+//     if we don't check `forwardedHost` and fail all pending requests, the requests with forwarding will be failed too. this may cause some issue:
+//     1. data race. see https://github.com/tikv/client-go/issues/1222 and TestRandomRestartStoreAndForwarding.
+//     2. panic which cause by `send on closed channel`, since failPendingRequests will close the entry.res channel,
+//     but in another batchRecvLoop goroutine,  it may receive the response from forwardedHost store2 and try to send the response to entry.res channel,
+//     then panic by send on closed channel.
+func (c *batchCommandsClient) failPendingRequests(err error, forwardedHost string) {
 	util.EvalFailpoint("panicInFailPendingRequests")
 	c.batched.Range(func(key, value interface{}) bool {
 		id, _ := key.(uint64)
 		entry, _ := value.(*batchCommandsEntry)
-		c.batched.Delete(id)
-		c.sent.Add(-1)
-		entry.error(err)
+		if entry.forwardedHost == forwardedHost {
+			c.failRequest(err, id, entry)
+		}
 		return true
 	})
 }
 
+// fastFailRequests fast fails requests by requestID.
+func (c *batchCommandsClient) fastFailRequests(err error, requestIDs []uint64) {
```

Comment:
How about to rename it to `failRequestsByIDs`?

Reply:
great, done.